StealFocus/steal-focus-explainer.md

```diff
@@ -0,0 +1,79 @@
+
+## Use Cases
+Some web apps such as Microsoft Teams apps are often run inside an iFrame in the web. This has resulted in some poor user experience issues for keyboard users. For example: in Teams it’s possible to view notifications of apps in the ‘Activity’ section. When a user navigates to a notification from an app (such as Viva Engage) that app can ‘steal focus’ from the user and place it in their app. This prevents the user from their keyboard to navigate through their notifications because an app can ‘steal’ focus at any time. Keyboard users' expectations are that their keyboard focus doesn’t jump around unexpectedly.
+
+## Proposed solution: "disallowprogrammaticfocus" attribute
+We propose a new "disallowprogrammaticfocus" boolean attribute on the [HTMLIFrameElement](https://www.w3.org/TR/2011/WD-html5-20110525/the-iframe-element.html#the-iframe-element). Whenever it is set, all the nested iframes can no longer take input focus through script.
+
+```html
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title>iframe steal focus prevention</title>
+  </head>
+  <body>
+    <iframe id="extension-tab-frame"
+            class="extension-tab-frame"
+            name="extension-tab-frame"
+            src=""
+            frameborder="0"
+            sandbox="allow-forms allow-modals allow-popups allow-pointer-lock allow-scripts allow-same-origin allow-downloads"
+            allow="geolocation *; microphone *; camera *; midi *; encrypted-media *;"
+            allowfullscreen
+            disallowprogrammaticfocus>
+    </iframe>
+  </body>
+</html>
+```
+
+
+
+## Alternative Solutions
+This section exposes some of the alternative solutions that we came across before coming up with the chosen proposal.
+
+### Add "programmaticfocus" option to "allow" or "sandbox" attribute.
+Similar to the "media-playback-while-not-rendered" [Permission Policy](https://www.w3.org/TR/permissions-policy/) on the [HTMLIFrameElement](https://www.w3.org/TR/2011/WD-html5-20110525/the-iframe-element.html#the-iframe-element). We chould have a new policy "programmatic-focus". When set, the nest iframes could take input focus through script.
+
+```html
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title>iframe steal focus prevention</title>
+  </head>
+  <body>
+    <iframe id="extension-tab-frame"
+            class="extension-tab-frame"
+            name="extension-tab-frame"
+            src=""
+            frameborder="0"
+            sandbox="allow-programmatic-focus allow-forms allow-modals allow-popups allow-pointer-lock allow-scripts allow-same-origin allow-downloads"
+            allow="programmatic-focus *; geolocation *; microphone *; camera *; midi *; encrypted-media *;"
+            allowfullscreen>
+    </iframe>
+  </body>
+</html>
+```
+
+This alternative was not selected as the preferred one, because we think that the default option (policy not set) is against current behavior and may have compatibility risk.
```

Comment:
Seems like it would be web compatible if we allowed this functionality to be controlled with a new permissions policy whose default value is '*'.

Not sure if there are any permissions policies with the fully permissive default like that. But if that's workable it could be a lighter weight approach versus a whole new attribute, and it could be easier for sites to adopt if they're already using permissions policies to control other things.

Reply:
I am not sure I get it. Are you suggesting we create a new policy like `disallow-programmatic-focus`?

```
<iframe src=""
        allow="disallow-programmatic-focus *;">
</iframe>
```

Reply:
In order to match the polarity of existing [permissions policies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Permissions_Policy), I think we'd want to have the polarity the other way around, so that denying the policy disables the functionality. So it would be a policy like `take-programmatic-focus`, with a default allowlist of `'*'` (for backwards compatibility). Then, to restrict it for iframes, a page could either be served with the HTTP header `Feature-Policy: take-programmatic-focus 'none'`, or use the `allow` attribute on the iframe:
`<iframe src="" allow="take-programmatic-focus none;"></iframe>`.

Per the email thread about this, it sounds like Ian may have proposed something along these lines?